Element Count Per Agent/Element Count Per Agent.cs

```diff
@@ -113,150 +119,238 @@ public void OnStopUpdates()
+                if (oldElementInfo.HostingAgentID != elementInfo.HostingAgentID)
+                {
+                    _elementInfoCache[elementId] = elementInfo;
 
-        private GetDataMinerInfoResponseMessage[] LoadAgents()
-        {
-            if (_dms == null)
-                throw new ArgumentNullException($"{nameof(GQIDMS)} is null.");
+                    if (!_rowCache.TryGetValue(oldElementInfo.HostingAgentID, out var oldHostRow))
+                    {
+                        _logger.Warning($"Could not find row data for agent {oldElementInfo.HostingAgentID} while processing swarm of element {elementId} {elementInfo.Name} to agent {elementInfo.HostingAgentID}");
+                        return;
+                    }
 
-            DMSMessage[] resp = null;
-            try
-            {
-                var req = new GetInfoMessage(InfoType.DataMinerInfo);
-                resp = _dms.SendMessages(req);
-            }
-            catch (Exception ex)
-            {
-                throw new DataMinerSecurityException($"Issue occurred in {nameof(ElementCountPerAgent)} when sending request {nameof(GetInfoMessage)}.{InfoType.DataMinerInfo}: {ex}", ex);
-            }
+                    // update new host
+                    currentHostRow.TotalElementCount++;
+                    if (elementInfo.IsSwarmable)
+                        currentHostRow.SwarmableElementCount++;
+                    else
+                        currentHostRow.NonSwarmableElementCount++;
 
-            if (resp == null || resp.Length == 0)
-                throw new Exception($"Response is null or empty");
+                    // but also update old host
+                    oldHostRow.TotalElementCount--;
+                    if (elementInfo.IsSwarmable)
+                        oldHostRow.SwarmableElementCount--;
+                    else
+                        oldHostRow.NonSwarmableElementCount--;
 
-            var dmaResponses = resp.OfType<GetDataMinerInfoResponseMessage>().ToArray();
-            if (dmaResponses.Length == 0)
-                throw new Exception($"{nameof(dmaResponses)} is empty");
+                    if (shouldUpdateGQI)
+                    {
+                        _logger.Information($"Updating host {oldHostRow.AgentID} with counts ({oldHostRow.NonSwarmableElementCount}/{oldHostRow.SwarmableElementCount}/{oldHostRow.TotalElementCount})");
+                        _updater.UpdateRow(oldHostRow.ToGQI());
+                        _logger.Information($"Updating host {currentHostRow.AgentID} with counts ({currentHostRow.NonSwarmableElementCount}/{currentHostRow.SwarmableElementCount}/{currentHostRow.TotalElementCount})");
+                        _updater.UpdateRow(currentHostRow.ToGQI());
+                    }
+                    return;
+                }
 
-            return dmaResponses;
+                // Case 4: Some other update we are not interested in, ignoring
```

Comment:
case 4: same host, swarmable flag changed

Reply:
I can live without. Swarmability of elements will not change that often & reloading the app is fast now :)